torch/lib/c10d/GlooDeviceFactory.cpp

```diff
@@ -0,0 +1,103 @@
+#include <c10d/GlooDeviceFactory.hpp>
+
+#include <c10/util/Exception.h>
+
+#if GLOO_HAVE_TRANSPORT_TCP
+#include <gloo/transport/tcp/device.h>
+#endif
+
+#if GLOO_HAVE_TRANSPORT_UV
+#include <gloo/transport/uv/device.h>
+#endif
+
+// On Linux, check that the tcp transport is available.
+#ifdef __linux__
+#if !GLOO_HAVE_TRANSPORT_TCP
+#error "Expected the tcp transport to be available on Linux."
+#endif
+#endif
+
+// On macOS, check that the uv transport is available.
+#ifdef __APPLE__
+#if !GLOO_HAVE_TRANSPORT_UV
+#error "Expected the uv transport to be available on macOS."
+#endif
+#endif
+
+namespace c10d {
+
+C10_DEFINE_SHARED_REGISTRY(
+    GlooDeviceRegistry,
+    ::gloo::transport::Device,
+    const std::string& /* interface */,
+    const std::string& /* hostname */);
+
+#if GLOO_HAVE_TRANSPORT_TCP
+static std::shared_ptr<::gloo::transport::Device> makeTCPDevice(
+    const std::string& interface,
+    const std::string& hostname) {
+  TORCH_CHECK(
+      !interface.empty() || !hostname.empty(),
+      "GlooDeviceFactory::makeTCPDevice(): interface or hostname "
+      "can't be empty");
+
+  ::gloo::transport::tcp::attr attr;
+  if (!interface.empty()) {
+    attr.iface = interface;
+  } else {
+    attr.hostname = hostname;
+  }
+  return ::gloo::transport::tcp::CreateDevice(attr);
+}
+
+C10_REGISTER_CREATOR(GlooDeviceRegistry, TCP, makeTCPDevice);
+#endif
+
+#if GLOO_HAVE_TRANSPORT_UV
+static std::shared_ptr<::gloo::transport::Device> makeUVDevice(
+    const std::string& interface,
+    const std::string& hostname) {
+  TORCH_CHECK(
+      !interface.empty() || !hostname.empty(),
+      "GlooDeviceFactory::makeUVDevice(): interface or hostname "
+      "can't be empty");
+
+  ::gloo::transport::uv::attr attr;
+  if (!interface.empty()) {
+    attr.iface = interface;
+  } else {
+    attr.hostname = hostname;
+  }
+  return ::gloo::transport::uv::CreateDevice(attr);
+}
+
+C10_REGISTER_CREATOR(GlooDeviceRegistry, UV, makeUVDevice);
+#endif
+
+std::shared_ptr<::gloo::transport::Device> GlooDeviceFactory::
+    makeDeviceForInterface(const std::string& interface) {
+#ifdef __linux__
+  return GlooDeviceRegistry()->Create("TCP", interface, "");
```

Comment:
I'm a bit confused here so please let me know if I'm misunderstanding this.

What's the point of using the registry if you always access a specific implementation of it? Shouldn't this arg always be the same? From the fb specific diff it's just using a copy of this file w/ a different implementation of this method and different linking so it doesn't seem like you're using the registry overrides? If we're going to go that way might as well just remove all the registry code. 

cc @pietern

Reply:
+1

This should simply iterate over the available constructors and either 1) pick the first one, or 2) if an env var is set (e.g. `GLOO_TRANSPORT_TYPE`) pick that one.

Then there is no need to override this function and all you need to do is add a compilation unit where additional implementations are registered.

Reply:
Agree with @pietern and @d4l3k I was imagining a registry with key and priority. Pick the highest priority transport that's registered unless overwritten by an env which is the key to one of the registered transport. 
On the other hand since this its decided to be all an internal gloo implementation stuff with no interface change. I am fine doing it in an iteration over this.

Reply:
@pietern Please correct me if I'm wrong. You are suggesting to check env in open source code?

Reply:
@satgera To achieve this - pick up higher priority registration unless overwritten by env, I think we need to have two registries? One for platformRegistry with priorities, another one for transportRegistry without priorities.

Reply:
To recap my thoughts a bit - 1) I'm going to hide any other specific logic from open source, and it's not necessarily to be a simple add-on or simple copy. For example, open source provides TCP and UV, but the stacked implementation provides TCP and TLS (no UV). So I didn't reuse the registry but implemented factory override. 2) env check is just for internal purpose to be introduced. Thus I isolated from open source by overriding in separate implementation. 3) Registry with priority is fine but will make TLS by default and enforce all customers do TLS. From my perspective, providing env configuration is preferred witch ends up with another question 4) Which transport are we going to default in the stacked implementation? We can discuss over that diff. 5) Given these thinking, at the first beginning I was also a bit confused why to use registry. Maybe we can get rid of this constrain to use registry.

Given all of the thoughts, I still think it makes more sense to have separate factory implementation with better isolation and flexibility.

Reply:
> @pietern Please correct me if I'm wrong. To just add additional registration in a separate source file, you are suggesting to check env in open source code?

Its fine the check env but the env should be to the effect GLOO_TRANSPORT_OVERRIDE = "tcp"/"uv"/"xx"